src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs

```diff
@@ -21080,6 +21098,297 @@ .maxstack  3
 }");
         }
 
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public void WithExprReference_WithCovariantReturns(bool emitRef)
+        {
+            var src = @"
+public record C
+{
+    public int X { get; init; }
+}
+public record D(int Y) : C;";
+            var comp = CreateCompilation(RuntimeUtilities.IsCoreClrRuntime ? src : new[] { src, IsExternalInitTypeDefinition }, targetFramework: TargetFramework.StandardLatest);
+            comp.VerifyDiagnostics();
+
+            var src2 = @"
+using System;
+class E
+{
+    public static void Main()
+    {
+        var c = new C() { X = 1 };
+        var c2 = c with { X = 2 };
```

Comment:
Could we extract a method for `with` on `C` and one on `D`, and verify IL just on that? That would make the difference more compact and clearer.

Reply:
@jcouv Sorry for missing this comment. I updated the test.

Reply:
Thanks. We probably no longer need to verify IL for `Main` (`CHelper` and `DHelper` are the important ones)